### PR TITLE
Fix to make taps on views outside parent bounds work on Android

### LINKED ISSUE #21455: [0.57.1] Android overflow doesn't register taps (touchable) in the "overflowed" area
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
```
 React Native Environment Info:
    System:
      OS: macOS High Sierra 10.13.6
      CPU: x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
      Memory: 62.65 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.9.0 - /usr/local/bin/node
      Yarn: 1.7.0 - ~/.npm-packages/bin/yarn
      npm: 6.0.1 - ~/.npm-packages/bin/npm
      Watchman: 4.7.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 11.4, macOS 10.13, tvOS 11.4, watchOS 4.3
      Android SDK:
        Build Tools: 21.1.2, 23.0.1, 23.0.2, 23.0.3, 24.0.1, 25.0.0, 25.0.1, 25.0.2, 25.0.3, 26.0.0, 26.0.1, 26.0.2, 26.0.3, 27.0.1, 27.0.2, 27.0.3, 28.0.2
        API Levels: 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28
    IDEs:
      Android Studio: 2.3 AI-162.4069837
      Xcode: 9.4.1/9F2000 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.5.0 => 16.5.0 
      react-native: 0.57.1 => 0.57.1 
```

## Description
The "new" overflow feature works great now except the fact that it doesn't register the taps on the outside area. That's **Android** only.

## Reproducible Demo
```
<View
        style={{
        width: 50,
        height: 50,
        borderColor: 'blue',
        borderWidth: 2,
        overflow: 'visible'
    }}
>
    <TouchableOpacity style={{ width: 100, height: 100, backgroundColor: '#f006' }}>
        <Text>some button</Text>
    </TouchableOpacity>
</View>
```

Tap inside the blue square - works. Tap on the red square but outside the blue one - it doesn't

Am I doing something wrong?  

Thank you 🤗

### LINKED ISSUE #27061: i have a problem with touchable is not working in android, when one view is overlapping on another view
1>  i have design where has 2 view, and one button between 2 view, when i press on button that button is not wotking in android.
2> Means that button view is overlapping on other view, so that it is not working. 
3> it is happening only in android, it works fine in IOS.
4>
Screenshot 2019-10-30 at 4 58 31 PM

Please find a solution asap, thank you.

### LINKED ISSUE #27232: React Native Android: An absolute positioned, TouchableOpacity outside of the bounds of its parent is not clickable
If I place a TouchableOpacity outside the bounds of its parent (using absolute positioning) only the portions of the TouchableOpacity which are inside the bounds of the parent are clickable. I think this is a bug because this can only be reproduced in Android, not iOS.
React Native version: 35.0.0

## Steps To Reproduce

Here's a minimal code sample:
`
<View style={{height: 60, width: '100%', zIndex: 1}}>
    <TouchableOpacity onPress={() => console.log('pressed')} style={{position: 'absolute', zIndex: 2, backgroundColor: 'yellow', width: 100, height: 50, top: 70}} />
</View>
`

### PR DESCRIPTION
**Pull request by hsource**
## Summary

By default, Views in React Native have `overflow: visible`. When a child view is outside of the parent view's boundaries, it's visible on Android, but not tappable. This behaviour is incorrect, and doesn't match iOS behaviour.

- Taps on Views outside the bounds of a parent with `overflow: visible` (or unset) should register
- Taps on Views outside the bounds of a parent with `overflow: hidden` should continue to not register

Related issues:

- fixes #21455
- fixes #27061
- fixes #27232

### Fix

- Made `findTouchTargetView` not check that the touch was in the bounds of the immediate children, but instead
  - Check that the touch is in its own bounds when returning itself
  - Check that the touch for a child is in its own bounds only when `overflow: hidden` is set
- Modified related code to adjust to this change
- Added RNTesterApp test

## Changelog

[Android] [Fixed] - Allow taps on views outside the bounds of a parent with `overflow: hidden`

## Test Plan

This can be tested with 2 examples added to the bottom of the PointerEvents page of the RNTesterApp:

| Before | After |
| --- | --- |
| Before | After |

### CODE REVIEW

**Comment by hsource** [comment]
Actually, there are some issues with this merge request on certain view hierarchies. I'll fix them and redo this.

**Comment by hsource** [comment]
> Actually, there are some issues with this merge request on certain view hierarchies. I'll fix them and redo this.

The issues were with ReactCompoundView and ReactCompoundViewGroups - I pushed a new change to fix them!

File: RNTester/js/examples/PointerEvents/PointerEventsExample.js [hunk 1]
```diff
@@ -191,6 +243,18 @@ const exampleClasses: Array<ExampleClass> = [
     description:
       "`box-only` causes touch events on the container's child components to pass through and will only detect touch events on the container itself.",
   },
+  {
+    Component: OverflowVisibleExample,
+    title: '`overflow: visible`',
+    description:
+      '`overflow: visible` style should allow subelements that are outside of the parent box to be touchable.',
```
Comment (pxpeterxu) [inline RNTester/js/examples/PointerEvents/PointerEventsExample.js:250]: To make the test a bit more obvious, you might want to just add:

> Tapping the parts of Box B outside Box A should print "B touched" and "A touched", and tapping Box C should also print "C touched" and "A touched"

Comment (hsource) [reply RNTester/js/examples/PointerEvents/PointerEventsExample.js:250]: Done!

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/TouchTargetHelper.java [hunk 2]
```diff
@@ -111,18 +127,21 @@ private static View findClosestReactAncestor(View view) {
    * be relative to the current viewGroup. When the method returns, it will contain the eventCoords
    * relative to the targetView found.
    */
-  private static View findTouchTargetView(float[] eventCoords, ViewGroup viewGroup) {
-    int childrenCount = viewGroup.getChildCount();
-    // Consider z-index when determining the touch target.
-    ReactZIndexedViewGroup zIndexedViewGroup =
+  private static View findTouchTargetView(
+      float[] eventCoords, View view, EnumSet<TouchTargetReturnType> allowReturnTouchTargetTypes) {
+    if (allowReturnTouchTargetTypes.contains(TouchTargetReturnType.CHILD)
```
[the remark this answers is not in the document]
Comment (hsource) [reply ReactAndroid/src/main/java/com/facebook/react/uimanager/TouchTargetHelper.java:132]: Done!

Added:
```java
    // We prefer returning a child, so we check for a child that can handle the touch first
    // We prefer returning a child, so we check for a child that can handle the touch first
    if (allowReturnTouchTargetTypes.contains(TouchTargetReturnType.CHILD)
```

```java
    // Check if parent can handle the touch after the children
    if (allowReturnTouchTargetTypes.contains(TouchTargetReturnType.SELF)
```

**Comment by Rhomennik** [comment]
+1

**Comment by mildfuzz** [comment]
I'd like to trial this locally, any tips on how I can point my local dev env at this fork?

**Comment by hsource** [comment]
@mildfuzz - you can fork react-native and install it from Github! Here's some docs I wrote for my company a while back. Note that it will take you probably at least half an hour or an hour to run through this:

# Patching React Native

## Why do it

Sometimes, something in React Native is broken. We can fix it directly instead of waiting for a patch.

Here are past examples:

1. [Fixed TextInput not being selectable in removeClippedSubviews FlatLists](https://github.com/facebook/react-native/pull/28852)
   - This bug was hard to work around, and a fix was easy. This is a prime example where patching React Native makes sense.
2. [Fixed italic style not being removed on TextInput on Android](https://github.com/facebook/react-native/pull/28868)
   - This bug was honestly easy to do a workaround for, but I was already fixing the first bug. In most cases, I would recommend just working around this bug.

## How-to

**Prerequisite:** You'll need write access to Harry's repo. Ask him for it!

### Part 1: Fixing the issue inside `react-native`

I recommend adding a basic test-case inside of the RNTesterApp so that you don't have to redo `yarn install` to test your fix for the issue.

1. Clone our forked React Native repo [hsource/react-native](https://github.com/hsource/react-native)
2. Build the RNTesterApp using [instructions](https://github.com/facebook/react-native/tree/master/RNTester) in the React Native repo
   - If working on Android and your build fails because of an error with `wrapNoInt`, consider applying [this change on the harry-fixed-gradle branch](https://github.com/hsource/react-native/commit/5fbc1a226ae30633d2b8cc21dc35bb53f1bb0d7e) and then rebuilding
3. Run the RNTesterApp and verify it works
4. Add an example or use an existing example to verify what's broken
5. Do your fix (in native code, or in React Native Javascript)
6. If doing your fix in native code, you'll have to rebuild the app to test your fix. Otherwise, you can just reload the Javascript to tester your fix

### Part 2: Submitting the change

We're going to cherry-pick your change into a branch that has all the other fixes we've applied over time.

1. Create a branch for your new change. Consider creating a pull request for the original React Native project too so that we can eventually just go back to the upstream implementation
2. Look in this project's `package.json` to see what tag of React Native we're on. It should be something like `v0.62.2-wanderlog.1`
3. Find the corresponding branch: given the tag above, the branch should be something like `0.62-wanderlog`. Checkout that branch: `git checkout 0.62-wanderlog`
4. Cherry-pick your change(s) into the branch of React Native and push the updated branch.

Putting it together:

```sh
git checkout 0.62-wanderlog
git cherry-pick MY_COMMIT
git push
```

### Part 3: Publishing a built version

We're going to update the "built" branch and create a new tag you can `yarn install`

1. Checkout the built version of the branch. If you'd checked out `0.62-wanderlog` before, you want to checkout `0.62-wanderlog-built`
2. Reset that branch to the new version of your pre-built branch with `git reset --hard 0.62-wanderlog`
3. Run `yarn install` to make sure you have the right versions of everything installed
4. Build the Android Java JARs by running `./gradlew :ReactAndroid:installArchives`
   - I figured this out by reading the publish-npm.sh file
   - Pre-requisites:
   - You may need to install Android Studio and NDK: see https://developer.android.com/studio/projects/install-ndk
   - Alternatively, you can install Docker and run the [Docker command found here](https://github.com/facebook/react-native/wiki/Building-from-source#publish-your-own-version-of-react-native)
5. Remove `/android/` from the .gitignore file
6. Update the version in package.json
7. Commit a new version of the branch
8. Push the latest version of all files

```sh
git checkout 0.62-wanderlog-built
git reset --hard 0.62-wanderlog

# OPTION 1:
# You may need to install Android Studio and NDK for the following
# command: see https://developer.android.com/studio/projects/install-ndk
./gradlew :ReactAndroid:installArchives

# OPTION 2:
# Install Docker and run the command here:
# (from https://github.com/facebook/react-native/wiki/Building-from-source#publish-your-own-version-of-react-native)
docker run --rm --name rn-build -v $PWD:/pwd -w /pwd reactnativecommunity/react-native-android /bin/sh -c "./gradlew :ReactAndroid:installArchives"

# Manually update the version in package.json
# Manually remove /android/ from .gitignore
git add .
git commit -m "[0.62.2-wanderlog.XXX] Added fix for something and built"
```

### Part 4: Create a tag and install the package using your tag

1. Go to the project on Github and create a new "Release". Name it "v0.62.2-wanderlog.2", making sure the tag's name matches the new version in package.json
2. In the React Native project, do `yarn add hsource/react-native#v0.62.2-wanderlog.2`
3. Built the code and verify your fix works, then commit it!
   - Try to test on both Android and iOS, just in case

**Comment by hsource** [comment]
@mildfuzz I have this change in a fork of v0.62.2 I have at hsource/react-native#v0.62.2-wanderlog.9 - you can install it from Github with `yarn add hsource/react-native#v0.62.2-wanderlog.9` if you want to use it for now! Note that I make no guarantees that this repo/tag will be available forever, so you might want to fork that repo yourself and install it

**Comment by Shuo-Mr** [comment]
May I ask if there will be a merger in the near future? This problem has troubled me for many days

**Comment by lgenzelis** [comment]
I hope this PR gets more attention. It's a very important fix! Having to work around this bug is a pain in the neck 😅

**Comment by JeremiahUL** [comment]
Is there any updates on this PR? This is very very important

**Comment by MrToxy** [comment]
Please merge soon

**Comment by wvfan** [comment]
When I tried to make a dropdown with scroll, I found that in the overflow area, the ScrollView will not response to the scroll action. For the same ScrollView, in the inside area it can be scrolled. Besides, in the overflow area, the touchstart, touchmove will only triggered on the last child instead of the whole element chain. It's not a big issue, but may be it's related to the ScrollView no response? Anyway, this is a great work and makes many things much easier.

**Comment by fendorio** [comment]
This should be a priority PR, I have encountered this issue on 3 apps that I have worked on in the past 2 years alone, evidently it is a common issue experienced by numerous other people.

**Comment by safaiyeh** [comment]
cc @hramos @JoshuaGross Could someone check out this PR? It’s been a while since it was open and resolves year old issues

**Comment by solo29** [comment]
please merge it!!!

**Comment by Theofilos-Chamalis** [comment]
I agree with the people that have left their comments previously, this is a quite old and critical bug that I have also encountered in 5 or so RN projects and some times have to work around with design changes.

**Comment by kelset** [comment]
Hey @hsource I can't do much from my side to help get this merged, I tried to put the link to it in front of the FB team, I hope it will help. In the meantime I'm sure it'd be great if you could resolve the conflict 🙏

**Comment by JoshuaGross** [comment]
I fixed some merge errors and have been testing it internally. Looks good overall, I think. But... One implication I'm not sure I understand or is intentional:

If View A has overflow: visible, and view C is completely outside the bounds of A (or you tap on a part of C that is outside of the bounds of A). I would expect to see a touch event for C only, and not for A - because the tap did NOT occur within the bounds of A. But I see:

```
 LOG  B overflowing touched
 LOG  A overflow visible touched
```

Is that intentional, correct, and/or consistent with the way iOS works? If that's how touch events work already on iOS and it's consistent with W3C then I'm fine with it, but it seems unintuitive to me.

Thanks for fixing this and being patient with us, @hsource <3

**Comment by hsource** [comment]
@JoshuaGross - thanks a ton for taking this on! I saw you already rebased - let me know if you want any more help with it.

Yep, that's how it works on web too! I made a quick [demo to demonstrate it](https://stackblitz.com/edit/react-ts-ndb6ts). This is because events by default bubble up ([Wikipedia article](https://en.wikipedia.org/wiki/Event_bubbling)) to the parents

Demo

**Comment by Cixelyn** [comment]
Dang @hsource , you beat me to the example by a few minutes hah. I had my [JS fiddle](https://jsfiddle.net/gcpw3ksy/) all ready too.

To add to the discussion, the official W3 rules are here: https://www.w3.org/TR/DOM-Level-2-Events/events.html#Events-flow-bubbling Once the event is generated, I believe it must bubbles upwards in the tree if it is designated as bubbling, regardless of parent bounding boxes at that point.

And to address @JoshuaGross's other point, the behavior is also consistent w/ the current iOS behavior; see the following snack: https://snack.expo.dev/T-ijRZIqA

IMG_3C853A770EA8-1

**Comment by JoshuaGross** [comment]
Thanks for the context! I'll do a bit more review and test internally.

**Comment by JoshuaGross** [comment]
Landing, should be merged shortly.

**Comment by safaiyeh** [comment]
This is awesome thanks @JoshuaGross !

**Comment by danielgindi** [comment]
Not in 0.65?

**Comment by kelset** [comment]
@danielgindi no, this commit was merged 12 days ago - the 0.65 branch was cut way before that. You can try requesting this commit as cherry pick for 0.65.1

**Comment by JakubKohoutek** [comment]
Hi @hsource, thank you for the PR, awesome work 🌈 I was testing the fix and for me inside ScrollView with `overflow: 'visible'`  on Android taps do work, but scrolling doesn't. It seems it behaves the same as taps before the fix (working only if touched in bounds of the parent element) - do you please have an idea on how to solve this one as well?

**Comment by hungvu193** [comment]
hey, how can I apply this patch into my current project which using 0.64.1 version of react native?

**Comment by renanbronchart** [comment]
I can't upgrade for now.
Is there any fix to use with React-native v0.65?

**Comment by jamesncampbell** [comment]
> hey, how can I apply this patch into my current project which using 0.64.1 version of react native?

@hungvu193 Did you figure out how to do this? I tried to patch it onto 0.65.2 but it's still not working for me

**Pull request closed without merging**